### ISSUE TITLE
Table of Contents for Jupyter Notebooks

### ISSUE THREAD

**Issue by collinmccarthy** [microsoft/vscode-jupyter#1348]
# Feature: Notebook Editor, Interactive Window, Python Editor cells

## Description
For large Jupyter notebooks having a table of contents is absolutely necessary, but a pain to keep updated yourself using HTML tags. Jupyter notebook and JupyterLab both support tables of contents through extensions. Is it possible for you to add support for this in VS Code? Without it many of us who use Jupyter for large scientific notebooks can't justify switching over to VS Code for writing notebooks, even though we love the environment.

Thank you!

**Comment by GitHunter0** [comment]
@rchiodo , I saw that you guys migrated to vscode-jupyter. That worries me because the functionality to use .py files as Notebooks just by simply adding cells via `# %%`  was awesome, it combined the best of both worlds (script and notebook).
Is that going to be lost? Correct me if I'm wrong please. Thank you

**Comment by halapano** [comment]
+1 for TOC in VS code

**Comment by rchiodo** [comment]
> @rchiodo , I saw that you guys migrated to vscode-jupyter. That worries me because the functionality to use .py files as Notebooks just by simply adding cells via # %% was awesome, it combined the best of both worlds (script and notebook).
Is that going to be lost? Correct me if I'm wrong please. Thank you

Our existing functionality (you're describing the interactive window) will continue to work for python. We plan to expand the interactive window to also work for other languages.

**Comment by roma-glushko** [comment]
ToC is a brilliant idea 🙌

**Comment by nicebear1999** [comment]
vote for the toc feature of jupyter in vscode

**Comment by nizarmasmoudi** [comment]
+1 on TOC. I prefer working with vscode rather than jupyterlab especially when managing a whole project and writing multiple python scripts aside from notebooks. Besides, I feel like it's easier to navigate if I have my workspace in a different window than my navigator where I usually do research. Anyway, I would really love an extension or update concerning TOC in vscode

**Comment by davidxcohen** [comment]
+1 on ToC

**Comment by DROMZEE** [comment]
+1 on ToC

**Comment by rchiodo** [comment]
As an FYI, this is present in the 'outline' view if you use VS code insiders:

![image](https://user-images.githubusercontent.com/19672699/104233804-599e8f80-5407-11eb-8994-934035ac1ca4.png)

You should be able to jump around to different markdown cells and code cells under those markdown cells. It's still a work in progress, but the basic functionality should be working.

**Comment by Bakerbakura** [comment]
Is there any idea when this outline view would be available in the main VS Code build?

**Comment by rchiodo** [comment]
Sorry, but that's likely months away. We're working on getting the new editor to have parity with the old one. Once that's ready we're going to turn on an experiment to try it for some small set of people. Assuming that goes well, then we'd slowly roll it out to more and more folks.

**Comment by hanmala123** [comment]
It's a great idea to present TOC in the outline view.
I really want to see TOC in jupyter extension.

**Comment by yassineAlouini** [comment]
That would be a neat feature. Thanks in advance. :)

**Comment by syagev** [comment]
If possible, I suggest to have this TOC available for `.py` files with percent markup. I guess for that you'll need to have the user select which kind of "Outline" view is of interest - the one for regular code files (which will not pickup e.g. markdown cells), or the one adjusted for notebooks.

My use case - in our team we manage a big `.py` file that acts as a sort of "catalog" of useful and relatively independent cells. This has to go under source control for version management and collaboration. The typical workflow is then to open an Interactive Window and execute a few of the cells of interest. So for finding these cells in the large file the described TOC is useful.

**Comment by GitHunter0** [comment]
@syagev , yes, I miss the TOC for .py files very much

**Comment by rchiodo** [comment]
@syagev can you enter a separate issue for TOC for python files? I believe this item is going to eventually just be closed when we have fully switched to the VS code notebook implementation but capturing TOC for python files sounds useful too.

**Comment by JanAndrosiuk** [comment]
+1 on ToC

**Comment by claudiaregio** [comment]
This is available today with Native Notebooks in [VS Code Insiders!](https://code.visualstudio.com/insiders/)
Download to try it out today :)

**Comment by XiaomoWu** [comment]
> This is available today with Native Notebooks in [VS Code Insiders!](https://code.visualstudio.com/insiders/)
> Download to try it out today :)

How does it differ from the existing outline view which is already implemented in both stable/insiders? I didn't notice any difference. Am I'm looking at the wrong place?

**Comment by Xinchengzelin** [comment]
Hope the issue to be closed soon!

**Comment by mrrobert424** [comment]
@claudiaregio. Thank you for the improvement! What I was hoping for was a table of contents with **numbering** that appeared in both the outline and the notebook. Perhaps this is possible and I just don't know how to do it. Apologies if my oversight.

**Comment by rchiodo** [comment]
> @claudiaregio. Thank you for the improvement! What I was hoping for was a table of contents with **numbering** that appeared in both the outline and the notebook. Perhaps this is possible and I just don't know how to do it. Apologies if my oversight.

Sorry but numbering is not automatic. You can enter a feature request on [vscode ](https://github.com/microsoft/vscode/issues) to add numbering, or you can kinda cheat and add them yourself. Note that any numbers rendered by the markdown renderer don't show up though. You have to add them in such a way that they're not considered a numbered list by the markdown renderer:

![image](https://user-images.githubusercontent.com/19672699/121052830-98930000-c76f-11eb-82a3-27447a4a4b34.png)

**Comment by claudiaregio** [comment]
@XiaomoWu If you are seeing this in both insiders and stable that means you have been opted into the Native Notebooks experiment so you are all set for now :)

**Comment by jd-santos** [comment]
Would numbered ToC need to be a separate request? Either way I might install Insiders just to try this out because it's one of my last "must-haves" for leaving Jupiter.

**Comment by rchiodo** [comment]
> Would numbered ToC need to be a separate request?

Yes it's not really supported. You should log it on [VS code](https://github.com/microsoft/vscode/issues) as the outline view is not created by the Jupyter extension.

**Comment by rleyvasal** [comment]
@claudiaregio, I do not see the Table of Contents (Markdown) under outline using the latest version of VS Code Insiders 1.59.0. 

How to make it work?
<img width="616" alt="toc" src="https://user-images.githubusercontent.com/61335136/126020921-3ba2cd63-31aa-4385-8ca7-e350f1e39cf7.PNG">

**Comment by eyalt** [comment]
> 
> 
> @claudiaregio, I do not see the Table of Contents (Markdown) under outline using the latest version of VS Code Insiders 1.59.0.
> 
> How to make it work?
> <img alt="toc" width="616" src="https://user-images.githubusercontent.com/61335136/126020921-3ba2cd63-31aa-4385-8ca7-e350f1e39cf7.PNG">

Best I can suggest is to download VS Code Insiders version. It has Native Notebooks and does show the outline

**Comment by rchiodo** [comment]
@rleyvasal what's your version of the jupyter extension? That UI you're showing is the old notebook UI. That shouldn't be possible in insiders anymore.

**Comment by claudiaregio** [comment]
@rleyvasal Do you have a setting that opts you out of all experiments?

**Comment by rleyvasal** [comment]
@claudiaregio. Thank you for your help!  it is working fine now. 

I did have the setting to opt out - I uninstall and re-installed VSCode Insiders and the previous Settings were still there.

**Comment by claudiaregio** [comment]
@Xinchengzelin Please file a new feature request https://github.com/microsoft/vscode
